etcd 3.6 client update

liggitt opened the pull request:
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

Update to etcd 3.6 libraries and drop cloud library dependencies

#### Which issue(s) this PR fixes:
Fixes https://github.com/kubernetes/kubernetes/issues/113366

#### Special notes for your reviewer:

```diff
-Direct Dependencies: 184 
-Transitive Dependencies: 327 
-Total Dependencies: 365 
+Direct Dependencies: 182 
+Transitive Dependencies: 209 
+Total Dependencies: 247 
```

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

/cc @dims
/area code-organization

/cc @serathius 
for visibility to how etcd 3.6 client library updates will help clean up k8s deps (lets us pick up https://github.com/etcd-io/etcd/pull/16595 🎉 )

/cc @princepereira
for visibility to hnslib use of opencensus.io pulling in super old grpc / genproto dep (xref https://github.com/kubernetes/kubernetes/pull/127713#discussion_r1821212093 - fixed in https://github.com/kubernetes/kubernetes/pull/128841)

liggitt committed:
Fix kubeadm etcd client interface

liggitt commented on hack/unwanted-dependencies.json:
```diff
@@ -201,13 +181,17 @@
         "sigs.k8s.io/kustomize/kyaml"
       ],
+      "github.com/klauspost/compress": [
+        "github.com/prometheus/client_golang",
```
looks like this is the only caller, and intends to switch to stdlib once a zstd impl is merged

* https://github.com/prometheus/client_golang/blob/main/prometheus/promhttp/http.go#L469
* https://github.com/golang/go/issues/62513 (proposal accepted, not yet implemented)

serathius commented:
cc @ahrtr

aojea commented:
/cc @aojea

pohly commented:
/remove-wg device-management

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/go.mod:
```diff
@@ -59,18 +59,18 @@ require (
 	github.com/jonboulle/clockwork v0.4.0 // indirect
 	github.com/josharian/intern v1.0.0 // indirect
 	github.com/json-iterator/go v1.1.12 // indirect
+	github.com/klauspost/compress v1.17.9 // indirect
```
looks like this is the only caller that requires us to vendor this, and intends to switch to stdlib once a zstd impl is merged

* https://github.com/prometheus/client_golang/blob/main/prometheus/promhttp/http.go#L469
* https://github.com/golang/go/issues/62513 (proposal accepted, not yet implemented)

liggitt commented on hack/unwanted-dependencies.json:
```diff
@@ -87,17 +87,8 @@
-      ],
-      "cloud.google.com/go/firestore": [
-        "google.golang.org/genproto"
+        "github.com/grpc-ecosystem/go-grpc-middleware/v2"
```
this is new and unexpected, need to chase this and eliminate it

liggitt replied:
already done upstream, bumped to v2.3.0 to resolve this

liggitt commented on hack/unwanted-dependencies.json:
```diff
@@ -252,22 +232,15 @@
       "golang.org/x/exp": [
         "github.com/antlr4-go/antlr/v4",
         "github.com/google/cel-go",
+        "github.com/grpc-ecosystem/go-grpc-middleware/v2",
```
this is new and unexpected, need to chase this and eliminate it

liggitt replied:
sorry, I should have been clearer, etcd's switch to `go-grpc-middleware/v2` was completely expected and desired

`go-grpc-middleware/v2` adding a dependency onto `golang.org/x/exp` is new and unexpected, and I need to go chase that in *that* dep

liggitt commented on hack/unwanted-dependencies.json:
```diff
@@ -252,22 +232,15 @@
-        "go.etcd.io/etcd/client/v3",
-        "go.etcd.io/etcd/server/v3"
+      "google.golang.org/appengine": [
+        "github.com/grpc-ecosystem/go-grpc-middleware/v2"
```
this is new and unexpected, need to chase this and eliminate it

liggitt replied:
already done upstream, bumped to v2.3.0 to resolve this

ahrtr commented on go.mod:
```diff
@@ -54,34 +54,34 @@ require (
-	go.etcd.io/etcd/api/v3 v3.5.16
-	go.etcd.io/etcd/client/pkg/v3 v3.5.16
-	go.etcd.io/etcd/client/v3 v3.5.16
+	go.etcd.io/etcd/api/v3 v3.6.0-alpha.0.0.20250326120537-27336678946c
```
We have already cut etcd `release-3.6` branch, please try tag `v3.6.0-rc.2` (we will release rc.3 tomorrow).

liggitt replied:
hrmm... seeing an error trying to use that tag:

```
go work sync
go: go.etcd.io/etcd/server/v3@v3.6.0-rc.2 requires
	go.etcd.io/etcd/client/v2@v2.306.0-rc.2: missing go.etcd.io/etcd/client/go.mod and .../v2/go.mod at revision client/v2.306.0-rc.2
```

ahrtr replied:
We changed the client/v2 into an internal module https://github.com/etcd-io/etcd/tree/main/client/internal/v2.  Did you see similar issue when use the latest commit in main branch? 

The client/v2 is only used by the v2discovery, which has already been deprecated. We will completely remove it in 3.7.

liggitt replied:
> Did you see similar issue when use the latest commit in main branch?

no, `main` worked

liggitt replied:
> We changed the client/v2 into an internal module

that seems to be breaking use of `go.etcd.io/etcd/server/v3@v3.6.0-rc.2`

liggitt replied:
for go.etcd.io/etcd/server/v3/embed for in-process unit / integration tests

ahrtr replied:
But I agree it's a generic problem, we need to sort it out regardless. 

I think we made a mistake to hide the client/v2 (in https://github.com/etcd-io/etcd/pull/15591), which is still used by etcd/server. 

Two possible solutions that I can think of for now:
- either revert the change (do not hide client/v2 in 3.6);
- decommission the v2discovery in release-3.6. But it may break the existing users which are still using the deprecated v2discovery. 

cc @serathius

liggitt replied:
was it known / intentional that go.etcd.io/etcd/server/v3/embed become unusable?

liggitt replied:
> either revert the change (do not hide client/v2 in 3.6)

given how close we are to 3.6, this is simpler / safer

> decommission the v2discovery in release-3.6. But it may break the existing users which are still using the deprecated v2discovery.

this seems good to do long-term, in an orderly way... maybe announce a target removal release, and give a way to opt-into disabling v2discovery in 3.6 or 3.7, and plan to drop in 3.7 or 3.8?

ahrtr replied:
We can easily reproduce this problem, 

It seems any applications, which are depending on the `go.etcd.io/etcd/server/v3`, are affected.  :( 

```
$ go get go.etcd.io/etcd/server/v3@v3.6.0-rc.2
go: downloading go.etcd.io/etcd/server/v3 v3.6.0-rc.2
go: go.etcd.io/etcd/server/v3@v3.6.0-rc.2 requires go.etcd.io/etcd/client/v2@v2.306.0-rc.2: missing go.etcd.io/etcd/client/go.mod and .../v2/go.mod at revision client/v2.306.0-rc.2

```

ahrtr replied:
Raised https://github.com/etcd-io/etcd/issues/19670

Another blocker for 3.6.0 :(

ahrtr replied:
> maybe announce a target removal release

We are planning to remove it in 3.7. All v2 related stuff will gone in 3.7

liggitt replied:
I'm very confused by https://github.com/etcd-io/etcd/blob/v3.6.0-rc.2/server/go.mod#L31 is the confusing line ... no such module exists 

https://github.com/etcd-io/etcd/tree/client/v2.306.0-rc.2/client has no go.mod or v2/go.mod file ... that's required for that module at that tag to exist

Probably best to hoist this into an etcd repo issue

ahrtr replied:
> I'm very confused by the dependency at https://github.com/etcd-io/etcd/blob/v3.6.0-rc.2/server/go.mod#L31 ... no such module exists:

It does exist. The problem is that it's internal, so it can't be used by module outside the repo

https://github.com/etcd-io/etcd/blob/27336678946c98bf272460e5dfc54909ecafc63e/client/internal/v2/go.mod#L1

liggitt replied:
`internal` in the *package* means things, but that module does not exist in a way Go can resolve

ahrtr replied:
sorry for the inconvenience. One more lesson learned.

liggitt replied:
all good, glad I happened to try to update this and you happened to mention the rc tags were available so we caught this before .0 :)

ahrtr replied:
> all good, glad I happened to try to update this and you happened to mention the rc tags were available so we caught this before .0 :)

We will fill the gap in test :) i.e. adding an example application and verify all the new dependencies/tags are resolvable

serathius replied:
Could we move `go.etcd.io/etcd/client/v2` into `go.etcd.io/etcd/server/internal/client/v2`?

ahrtr replied:
> Could we move `go.etcd.io/etcd/client/v2` into `go.etcd.io/etcd/server/internal/client/v2`?

Seems also a valid option. Let's try it out.

liggitt replied:
+1, dropping the client/v2 dep entirely sounds really good

ahrtr replied:
Just raised https://github.com/etcd-io/etcd/pull/19673

liggitt replied:
updated to release-3.6 successfully

ahrtr replied:
thx for the confirmation.

liggitt commented:
unit test failures:
* cel compat test is due to golang.org/x/text bump, picked in test fixup commit from https://github.com/kubernetes/kubernetes/pull/130913 to address

@serathius, the TestWatchStreamSeparation and TestProgressNotify failures look like they're due to the etcd 3.6 bump... can you dig into those and see if that's an issue etcd needs to fix, or if we need to adjust our embedded server usage? If our usage is incorrect, it's surprising only those two tests fail... they're testing less commonly used features of etcd, so I wonder if there's an issue on the etcd side

ahrtr commented:
> Did something in the etcd/embed server slow down between 3.5 and 3.6?

We fixed a race condition when shutting down the embed etcdserver.  It may only slow down the shutdown process a bit. Refer to https://github.com/etcd-io/etcd/pull/19221 and https://github.com/etcd-io/etcd/pull/19139. The fixes were also backported to 3.5.18.  I believe K8s master branch is on an old etcd version, so you see the difference.

ahrtr commented:
> Hmm... if I _just_ run TestWatchStreamSeparation locally, the test passes:

Do you mean running it alone passes, but running it in workflow fails? Not sure if it's related to https://github.com/etcd-io/etcd/pull/19195#issuecomment-2627486682

liggitt commented:
on master:

```
go test -race k8s.io/apiserver/pkg/storage/etcd3 -count 1 -run TestProgressNotify
ok  	k8s.io/apiserver/pkg/storage/etcd3	3.448s
```

on this branch with etcd ~v3.6.0-rc.2:
```
go test -race k8s.io/apiserver/pkg/storage/etcd3 -count 1 -run TestProgressNotify
--- FAIL: TestProgressNotify (37.52s)
    utils.go:171: time out after waiting 30s on ResultChan
    logger.go:146: 2025-03-27T11:24:54.660-0400	ERROR	etcd-server	setting up serving from embedded etcd failed.	{"error": "http: Server closed"}
    logger.go:146: 2025-03-27T11:24:54.661-0400	ERROR	etcd-server	setting up serving from embedded etcd failed.	{"error": "mux: server closed"}
    logger.go:146: 2025-03-27T11:25:01.667-0400	ERROR	etcd-server	setting up serving from embedded etcd failed.	{"error": "accept tcp 127.0.0.1:65272: use of closed network connection"}
    logger.go:146: 2025-03-27T11:25:01.675-0400	ERROR	etcd-server	setting up serving from embedded etcd failed.	{"error": "accept tcp 127.0.0.1:65273: use of closed network connection"}
FAIL
FAIL	k8s.io/apiserver/pkg/storage/etcd3	38.384s
```

Looks like that test set clusterConfig.ExperimentalWatchProgressNotifyInterval and that is no longer effective... setting clusterConfig.WatchProgressNotifyInterval as well makes the test pass. Was it expected that ExperimentalWatchProgressNotifyInterval is non-functional in 3.6? If so, that's fine, it's easy to also set the non-experimental field.

liggitt commented:
> Do you mean running it alone passes, but running it in workflow fails?

That callout in the CI failure was a red herring... the issue was that the package timed out because it ran slightly slower. I don't think that's an etcd issue, that package hovers at the timeout boundary anyway.

ahrtr commented:
> Looks like that test set clusterConfig.ExperimentalWatchProgressNotifyInterval and that is no longer effective... setting clusterConfig.WatchProgressNotifyInterval as well makes the test pass. Was it expected that ExperimentalWatchProgressNotifyInterval is non-functional in 3.6? If so, that's fine, it's easy to also set the non-experimental field.

Oh, it's a known issue:) It's a breaking change for embedded use case, sorry. We are aware of it.  Refer to https://github.com/etcd-io/etcd/issues/18023#issuecomment-2653717397

Please either use the non-experimental flag or use feature-gate.

liggitt committed:
Set non-experimental WatchProgressNotifyInterval config field

ahrtr commented:
> Was it expected that ExperimentalWatchProgressNotifyInterval is non-functional in 3.6?

To be clearer, it only affects the embedded use cases. If you run etcd binnary or image, they are not affected.

ahrtr commented on staging/src/k8s.io/apiserver/pkg/storage/etcd3/watcher_test.go:
```diff
@@ -107,6 +107,7 @@ func TestWatchInitializationSignal(t *testing.T) {
 func TestProgressNotify(t *testing.T) {
 	clusterConfig := testserver.NewTestConfig(t)
 	clusterConfig.ExperimentalWatchProgressNotifyInterval = time.Second
```
You don't need to keep this line anymore.

liggitt commented:
Looks like the only failing test is timeout of k8s.io/apiserver/pkg/storage/cacher, due to slightly longer run times changing only the etcd 3.6.0-rc.3 dependency

The unit tests in this package have hovered around problematic timeout levels for a while (https://github.com/kubernetes/kubernetes/issues/123685, https://github.com/kubernetes/kubernetes/issues/123850, https://github.com/kubernetes/kubernetes/issues/125688).

@serathius, looks like the runtime of that package crept up slightly from 1.32 → 1.33

`go test -race k8s.io/apiserver/pkg/storage/cacher` results:
```
k8s.io/apiserver/pkg/storage/cacher	188.628s (this PR, which is master + etcd 3.6.0-rc.3)
k8s.io/apiserver/pkg/storage/cacher	166.207s (master, which is on etcd 3.5.16)
k8s.io/apiserver/pkg/storage/cacher	160.975s (release-1.32, which is on etcd 3.5.16)
```

Before we can bump kubernetes to etcd 3.6 in 1.34, we'll need to resolve this. Some possibilities:
* speed up tests (parallelize, or optimize some of the tests)
* slim down tests (if there's some that are redundant or could be consolidated or are no longer needed)
* split up tests (could consider splitting tests into multiple packages just to work around the 3 minute limit, though I'd prefer speeding up or slimming down if possible)

serathius commented:
> @serathius, looks like the runtime of that package crept up slightly from 1.32 → 1.33

Yea, that's my fault. I added a bunch of tests hoping we will fit in timeout. Just confirming, should we fix this after thaw? or we can slip in during test freeze?

liggitt commented:
After thaw is fine, just wanted to get it on your radar. If we have a change ready to go that resolves this by then, that would be super helpful to get the 3.6 bump landed first thing in 1.34

ahrtr commented:
> that would be super helpful to get the 3.6 bump landed first thing in 1.34

I think we need to land etcd 3.5.20 (or higher patch, 3.5.21 is the latest one for now) into 1.33.1 first, then 3.6.0 to 1.34 later, otherwise the upgrade from 3.5 to 3.6 may fail. cc @neolit123

liggitt commented:
That shouldn't block merging the 3.6 client into 1.34 though, right?

ahrtr commented:
> That shouldn't block merging the 3.6 client into 1.34 though, right?

Right, only the etcd image/binary version matters regarding the upgrade issue.

Client sdk and image doesn't necessarily to be bound together. 
Ideally, I'd see landing both 3.5.20+ client & image into 1.33 first, and 3.6.0 client & image into 1.34.

liggitt commented:
> Ideally, I'd see landing both 3.5.20+ client & image into 1.33.1 first, and 3.6.0 client & image into 1.34 later.

uh... if there are correctness bugs we want to be *sure* are fixed in the versions used in 1.33.x, we should be trying to land that in 1.33.0, not waiting for 1.33.1

ahrtr commented:
> if there are correctness bugs we want to be _sure_ are fixed in the versions used in 1.33.x,

As mentioned in https://etcd.io/blog/2025/upgrade_from_3.5_to_3.6_issue/, we fixed an upgrade issue (3.5 -> 3.6) in 3.5.20, and the root cause is the membership data maybe inconsistent between v2store (deprecated, but remain source of truth for membership data in 3.5) and v3store (source of truth in 3.6).

We also fixed a [Missing delete event on watch](https://github.com/etcd-io/etcd/issues/19179) in 3.5.18.

But all the above two fixes are only etcdserver side, no changes on etcd client SDK.



> we should be trying to land that in 1.33.0, not waiting for 1.33.1. Otherwise, upgrades using kubeadm 1.33.0 to 1.34.x will be risky, right?

Yes, it's correct. I was trying to catch up the code freeze to land 3.5.20 in 1.33, but still late one day as mentioned in https://github.com/etcd-io/etcd/issues/19557#issuecomment-2739815296. So we should apply for an exception?

liggitt commented:
> Yes, it's correct. I was trying to catch up the code freeze to land 3.5.20 in 1.33, but still late one day as mentioned in [etcd-io/etcd#19557 (comment)](https://github.com/etcd-io/etcd/issues/19557#issuecomment-2739815296). So we should apply for an exception?

Yes, I would think so. A bugfix we would consider valid for backport to release-1.33 is also valid to merge during code freeze before release, if we can do so confidently. Especially if it has implications for 1.33 → 1.34 upgrades with the etcd version kubeadm would want to use.

liggitt commented:
rebased on master, cacher unit tests are passing now with the speedups in https://github.com/kubernetes/kubernetes/pull/131103

liggitt commented:
update to v3.6.0-rc.3 tag went smoothly, all required CI is green now 🎉

ahrtr commented:
@liggitt is this PR still WIP (I see "WIP" is still in the title)? I think it's ready to merge, since release-1.33 branch has already cut.

liggitt commented:
I'll drop the WIP when I update to 3.6.0 tags

ahrtr commented:
https://github.com/etcd-io/etcd/releases/tag/v3.6.0-rc.5

liggitt commented:
updated with rc.5

ahrtr commented:
> updated with rc.5

thx

liggitt committed:
bump etcd client to 3.6

hack/pin-dependency.sh go.etcd.io/etcd/api/v3 v3.6.0
hack/pin-dependency.sh go.etcd.io/etcd/client/pkg/v3 v3.6.0
hack/pin-dependency.sh go.etcd.io/etcd/client/v3 v3.6.0
hack/pin-dependency.sh go.etcd.io/etcd/pkg/v3 v3.6.0
hack/pin-dependency.sh go.etcd.io/etcd/server/v3 v3.6.0

hack/pin-dependency.sh github.com/grpc-ecosystem/go-grpc-middleware/v2 v2.3.0

hack/update-vendor.sh

liggitt commented:
/hold cancel

etcd 3.6.0 is released, this is updated with tagged versions of the client libraries

liggitt commented on hack/unwanted-dependencies.json:
```diff
@@ -253,22 +228,12 @@
+        "github.com/grpc-ecosystem/go-grpc-middleware/v2",
         "k8s.io/mount-utils"
       ],
-      "google.golang.org/genproto": [
```
🎉

dims replied:
Sweet!

dims commented:
/approve
/lgtm

dims commented:
https://github.com/kubernetes/kubernetes/issues/128583 should close out with this PR as well

dims commented:
/skip

ahrtr commented:
/lgtm


Thanks!!!

aojea commented:
this is so good 😭
